### PR TITLE
CORE-5258 logging improvements to corda cli

### PR DESCRIPTION
**Pull request by pmck91**
Several changes to how the CLI write logs and output

We now capture all of stdOut and stdErr and log it to a file, as well as to the console. 
There is now a rolling log file located in the '~/.corda/cli/logs' directory, along with the rest of the install

These will be seen as coming from [System Out] or [System Error] in the log file to differentiate it from regular logging.

You can now supply -DlogFile and -DlogLevel to create another logfile that defaults to INFO if you do not supply logLevel to the jvm. This will be useful for e2e tests dealing with the plugins. 

A doc is to follow to describe all of the new changes as developers should consider using logger.info .error .warn etc instead of System.out.println()

### CODE REVIEW

File: app/src/main/kotlin/net/corda/cli/application/logger/LoggerStream.kt [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            val errOut: Logger = LoggerFactory.getLogger("System Err")
+
+            System.setOut(PrintStream(LoggerStream(sysOut, LogLevel.INFO)))
+            System.setErr(PrintStream(LoggerStream(errOut, LogLevel.ERROR)))
+        }
+    }
+
+    private val baos = ByteArrayOutputStream(1000)
+
+    override fun write(b: Int) {
```
Comment (benmcmahon100) [inline app/src/main/kotlin/net/corda/cli/application/logger/LoggerStream.kt]: What's with the variable names? Can we change them to something more descriptive for readability?

**Review by benmcmahon100 — CHANGES_REQUESTED** [review]

File: app/build.gradle [hunk 2]
```diff
@@ -21,7 +24,9 @@ dependencies {
     implementation "org.apache.servicemix.bundles:org.apache.servicemix.bundles.kafka-clients:$kafkaClientVersion"
 
     implementation "org.slf4j:slf4j-api:$slf4jVersion"
-    implementation "org.pf4j:pf4j:${pf4jVersion}"
+    implementation ("org.pf4j:pf4j:${pf4jVersion}") {
+        exclude group:"org.slf4j", module:"slf4j-api"
```
Comment (vkolomeyko) [inline app/build.gradle]: Any particular reason why we had to exclude `slf4j-api`, please ?

File: app/src/main/kotlin/net/corda/cli/application/logger/LogToFileAppender.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.core.LoggerContext
+import org.apache.logging.log4j.core.appender.FileAppender
+import org.apache.logging.log4j.core.layout.PatternLayout
+import java.io.File
+
+
+class LogToFileAppender {
+    companion object {
+        fun enableLogToFile(filePath: File, level: String) {
```
Comment (vkolomeyko) [inline app/src/main/kotlin/net/corda/cli/application/logger/LogToFileAppender.kt]: I think we are somewhat duplicating what we have in `log4j2.xml`.
If we need to control log level and location of the file, I think we should be able to do that directly in `log4j2.xml` like so:
```
<Root level="${env:LOG_LEVEL:-info}">
```
i.e. reading system properties directly.

Also, when we are adding an `FileAppender` here programmatically, are we sure we are not going to end-up with **3** appenders in this case:
1. Stdout;
2. Rolling file appender from `log4j2.xml`; 
3. `logToFileAppender` configured here
?

My preference would be to get rid of this class and have all the logging configuration settings in `log4j2.xml`.

File: app/src/main/kotlin/net/corda/cli/application/logger/LoggerStream.kt [hunk 4]
```diff
@@ -0,0 +1,68 @@
+package net.corda.cli.application.logger
+
+import java.io.OutputStream
+import java.io.PrintStream
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import java.io.ByteArrayOutputStream
+
+enum class LogLevel {
+    TRACE, DEBUG, INFO, WARN, ERROR
```
Comment (vkolomeyko) [inline app/src/main/kotlin/net/corda/cli/application/logger/LoggerStream.kt]:
```suggestion
    INFO, ERROR
```
Given that we only ever create `LoggerStream` with `INFO` and `ERROR`...

File: app/src/main/resources/log4j2.xml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        </File>
+
+    </Appenders>
+
+    <Loggers>
+        <!-- avoid duplicated logs with additivity=false -->
+        <Logger name="org.pf4j" level="info" additivity="false">
+            <AppenderRef ref="LogToFile"/>
+        </Logger>
+        <Root level="info">
```
Comment (vkolomeyko) [inline app/src/main/resources/log4j2.xml]: I think the plan was to allow overriding the log level via `-D` parameter:
```suggestion
        <Root level="${sys:logLevel:-info}">
```

File: app/src/main/kotlin/net/corda/cli/application/logger/LoggerStream.kt [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+class LoggerStream(
+    private val logger: Logger,
+    private val logLevel: LogLevel
+) : OutputStream() {
+
+    public companion object {
+        fun redirectSystemAndErrorOut() {
+            val sysOut: Logger = LoggerFactory.getLogger("System Out")
```
Comment (vkolomeyko) [inline app/src/main/kotlin/net/corda/cli/application/logger/LoggerStream.kt]: As for Plugin Developer experience, could we provide some guidance (possibly in the ReadMe) on how to perform logging correcytly from within a plugin?

I.e. what would developer has to do to output a string to a CLI console?
(Do we encourage use of `System.out.println()`? Or should the be getting hold of a named logger to perform the operation?) 
What needs to do output a message that will go into a file but not to CLI console? (like some internal log line)

File: app/src/main/resources/log4j2.xml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+            <AppenderRef ref="LogToConsole"/>
```
Comment (vkolomeyko) [inline app/src/main/resources/log4j2.xml:27]: If I do:
```
LoggerFactory.getLogger(MyClass.class).info("My message")
```
I would expect this log line to be only sent to a file but not to CLI console.
What would prevent this log line for appearing on CLI console, please?

File: app/build.gradle [hunk 8]
```diff
@@ -21,7 +24,7 @@ dependencies {
     implementation "org.apache.servicemix.bundles:org.apache.servicemix.bundles.kafka-clients:$kafkaClientVersion"
 
     implementation "org.slf4j:slf4j-api:$slf4jVersion"
-    implementation "org.pf4j:pf4j:${pf4jVersion}"
+    implementation ("org.pf4j:pf4j:${pf4jVersion}")
```
Comment (vkolomeyko) [inline app/build.gradle]: We can probably revert this change:
```suggestion
    implementation "org.pf4j:pf4j:${pf4jVersion}"
```

File: app/src/main/kotlin/net/corda/cli/application/logger/LoggerStream.kt (hunk 6 above)
Comment (pmck91) [reply app/src/main/kotlin/net/corda/cli/application/logger/LoggerStream.kt]: I actually have another task CORE-6297 for updating the readme and docs. I would suggest developers use the named loggers, but at least we catch system out in case they dont. As for what shouldnt be in the console, we would have to set up another named logger. Since console is set to info by default.

File: app/src/main/resources/log4j2.xml (hunk 7 above)
Comment (pmck91) [reply app/src/main/resources/log4j2.xml:27]: that will end up on the console too. Maybe the way that system out and err are captured needs to be looked at so we can avoid this.

Comment (vkolomeyko) [reply app/src/main/resources/log4j2.xml:27]: I think we can remove `LogToConsole` from `Root` and add two more loggers like so:
```
<Logger name="System Out" level="info" additivity="false">
            <AppenderRef ref="LogToConsole"/>
            <AppenderRef ref="LogToFile"/>
</Logger>

<Logger name="System Err" level="info" additivity="false">
            <AppenderRef ref="LogToConsole"/>
            <AppenderRef ref="LogToFile"/>
</Logger>
```
This should ensure that only asomething logged using `System Out` and `System Err` gets written to CLI console. But we will always log everything into a file.

Comment (pmck91) [reply app/src/main/resources/log4j2.xml:27]: for some reason log4j wouldnt respect LogToFile if the level was set to the systems, I have updated the code now, but do we want to always log every system.out.println to the file too?

Comment (vkolomeyko) [reply app/src/main/resources/log4j2.xml:27]: > do we want to always log every system.out.println to the file too?

Yes, that would make sense to me. If anything is going wrong with CLI command - we can ask for **1** log file which will have everything we need: stdOut, stdErr and log lines from all the other logger along with timestamps. So we would be able to establish exactly what went on and in which sequence.

File: app/src/main/kotlin/net/corda/cli/application/logger/LoggerStream.kt [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+import java.io.PrintStream
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import java.io.ByteArrayOutputStream
+
+enum class LogLevel {
+    INFO, ERROR
+}
+
+class LoggerStream(
```
Comment (vkolomeyko) [inline app/src/main/kotlin/net/corda/cli/application/logger/LoggerStream.kt:13]:
```suggestion
internal class LoggerStream(
```

also I would move `sysOut` and `errOut` up and outside of the class such that they will be accessible by plugins as: `net.corda.cli.application.logger.sysOut`.

File: app/src/main/resources/log4j2.xml [hunk 10]
```diff
@@ -0,0 +1,39 @@
+<?xml version='1.0' encoding='UTF-8' ?>
+
+<configuration>
+
+    <Properties>
+        <!--        <Property name="logFile">"${sys:user.home}/.corda/cli/logs/corda-cli.log"</Property>-->
```
Comment (vkolomeyko) [inline app/src/main/resources/log4j2.xml]: We probably should remove this comment...

File: app/src/main/resources/log4j2.xml (hunk 7 above)
Comment (pmck91) [reply app/src/main/resources/log4j2.xml:27]: done

File: app/src/main/kotlin/net/corda/cli/application/logger/LoggerStream.kt (hunk 9 above)
Comment (pmck91) [reply app/src/main/kotlin/net/corda/cli/application/logger/LoggerStream.kt:13]: they wouldn't be accessible unless they were injected into the plugins via interface in the api module, however they can call LoggerFactory.getLogger('systemOut') etc once the plugin has started.

Comment (vkolomeyko) [reply app/src/main/kotlin/net/corda/cli/application/logger/LoggerStream.kt:13]: Ah, yes - I have forgotten about `api` aspect.
IMO accessing them using `LoggerFactory.getLogger('systemOut')` is fine.

**Review by vkolomeyko — APPROVED** [review]
LGTM

**Review by benmcmahon100 — APPROVED** [review]
